crates/collab/src/api/extensions.rs

```diff
@@ -0,0 +1,235 @@
+use crate::{
+    db::{ExtensionMetadata, NewExtensionVersion},
+    executor::Executor,
+    AppState, Error, Result,
+};
+use anyhow::{anyhow, Context as _};
+use aws_sdk_s3::presigning::PresigningConfig;
+use axum::{
+    extract::{Path, Query},
+    response::Redirect,
+    routing::get,
+    Extension, Json, Router,
+};
+use collections::HashMap;
+use hyper::StatusCode;
+use serde::{Deserialize, Serialize};
+use std::{sync::Arc, time::Duration};
+use time::PrimitiveDateTime;
+use util::ResultExt;
+
+pub fn router() -> Router {
+    Router::new()
+        .route("/extensions", get(get_extensions))
+        .route(
+            "/extensions/:extension_id/:version/download",
+            get(download_extension),
+        )
+}
+
+#[derive(Debug, Deserialize)]
+struct GetExtensionsParams {
+    filter: Option<String>,
+}
+
+#[derive(Debug, Deserialize)]
+struct DownloadExtensionParams {
+    extension_id: String,
+    version: String,
+}
+
+#[derive(Debug, Serialize)]
+struct GetExtensionsResponse {
+    pub data: Vec<ExtensionMetadata>,
+}
+
+#[derive(Deserialize)]
+struct ExtensionManifest {
+    name: String,
+    version: String,
+    description: Option<String>,
+    authors: Vec<String>,
+    repository: String,
+}
+
+async fn get_extensions(
+    Extension(app): Extension<Arc<AppState>>,
+    Query(params): Query<GetExtensionsParams>,
+) -> Result<Json<GetExtensionsResponse>> {
+    let extensions = app.db.get_extensions(params.filter.as_deref(), 30).await?;
+    Ok(Json(GetExtensionsResponse { data: extensions }))
```

Comment:
This is a good point, but isn't something we typically prioritize in the codebase currently. A lot of the DB calls currently return structs that are directly consumed by the protobuf or HTTP APIs.

It's tightly coupled, but it has worked for us so far because for most database methods, there is only one consumer, so it isn't a problem to tailor the method's return type to suit that one consumer.